File: packages/aura_sdk_v1_ts/src/symmetricEnvelopeV1.ts
import { createCipheriv, createDecipheriv, createHash, randomBytes } from "node:crypto";

import {
  deriveSessionPublicKeyV1,
  deriveSessionSymmetricKeyV1,
  deriveSharedSecretV1,
  type SessionPublicKeyV1,
  type SessionSecretKeyV1,
} from "./sessionKeyV1.ts";
import {
  encodeSessionEncryptionContextV1,
  validateSessionEncryptionContextV1,
  type AuraSessionEncryptionContextV1,
} from "./sessionEncryptionContextV1.ts";
import {
  encodeStormEncryptionBindingV1,
  validateStormEncryptionBindingV1,
  type StormEncryptionBindingV1,
} from "./stormEncryptionBindingV1.ts";

export const ENCRYPTED_ENVELOPE_V1_VERSION = 0x01;
export const ENCRYPTED_ENVELOPE_V1_ALGORITHM_ID = 0x01;
export const ENCRYPTED_ENVELOPE_V1_NONCE_LEN = 12;
export const ENCRYPTED_ENVELOPE_V1_TAG_LEN = 16;
export const AURA_SESSION_ENCRYPTION_AAD_CONTEXT_HASH_V1_DOMAIN_SEPARATOR = new TextEncoder().encode(
  "AURA_SESSION_ENCRYPTION_AAD_CONTEXT_HASH_V1",
);

export type AuraEncryptedEnvelopeV1 = {
  version: number;
  algorithmId: number;
  senderPublicKey: SessionPublicKeyV1;
  receiverPublicKey: SessionPublicKeyV1;
  nonce: Uint8Array;
  aadContextHash: Uint8Array;
  ciphertext: Uint8Array;
  sessionKeyId: Uint8Array;
};

export function encryptPayloadV1(
  senderSecretKey: SessionSecretKeyV1,
  receiverPublicKey: SessionPublicKeyV1,
  sessionEncryptionContext: AuraSessionEncryptionContextV1,
  stormEncryptionBinding: StormEncryptionBindingV1,
  nonce: Uint8Array,
  plaintext: Uint8Array,
): Uint8Array {
  const sharedSecret = deriveSharedSecretV1(senderSecretKey, receiverPublicKey);
  const sessionKey = deriveSessionSymmetricKeyV1({
    sharedSecret,
    sessionEncryptionContext,
    stormEncryptionBinding,
  });
  const aadMaterial = buildAadMaterialV1(sessionEncryptionContext, stormEncryptionBinding);
  const cipher = createCipheriv(
    "chacha20-poly1305",
    Buffer.from(sessionKey.bytes),
    Buffer.from(requireNonceV1(nonce)),
    { authTagLength: ENCRYPTED_ENVELOPE_V1_TAG_LEN },
  );
  cipher.setAAD(Buffer.from(aadMaterial));

  const ciphertext = Buffer.concat([
    cipher.update(Buffer.from(cloneBytesV1(plaintext, "plaintext"))),
    cipher.final(),
    cipher.getAuthTag(),
  ]);
  return new Uint8Array(ciphertext);
}

export function decryptPayloadV1(
  receiverSecretKey: SessionSecretKeyV1,
  senderPublicKey: SessionPublicKeyV1,
  sessionEncryptionContext: AuraSessionEncryptionContextV1,
  stormEncryptionBinding: StormEncryptionBindingV1,
  nonce: Uint8Array,
  ciphertext: Uint8Array,
): Uint8Array {
  const payload = cloneBytesV1(ciphertext, "ciphertext");
  if (payload.length < ENCRYPTED_ENVELOPE_V1_TAG_LEN) {
    throw new TypeError(
      `ciphertext must be at least ${ENCRYPTED_ENVELOPE_V1_TAG_LEN} bytes to include the AEAD tag`,
    );
  }

  const sharedSecret = deriveSharedSecretV1(receiverSecretKey, senderPublicKey);
  const sessionKey = deriveSessionSymmetricKeyV1({
    sharedSecret,
    sessionEncryptionContext,
    stormEncryptionBinding,
  });
  const aadMaterial = buildAadMaterialV1(sessionEncryptionContext, stormEncryptionBinding);
  const decipher = createDecipheriv(
    "chacha20-poly1305",
    Buffer.from(sessionKey.bytes),
    Buffer.from(requireNonceV1(nonce)),
    { authTagLength: ENCRYPTED_ENVELOPE_V1_TAG_LEN },
  );
  decipher.setAAD(Buffer.from(aadMaterial));
  decipher.setAuthTag(Buffer.from(payload.subarray(payload.length - ENCRYPTED_ENVELOPE_V1_TAG_LEN)));

  const plaintext = Buffer.concat([
    decipher.update(Buffer.from(payload.subarray(0, payload.length - ENCRYPTED_ENVELOPE_V1_TAG_LEN))),
    decipher.final(),
  ]);
  return new Uint8Array(plaintext);
}

export function buildEncryptedEnvelopeV1(
  senderSecretKey: SessionSecretKeyV1,
  receiverPublicKey: SessionPublicKeyV1,
  sessionEncryptionContext: AuraSessionEncryptionContextV1,
  stormEncryptionBinding: StormEncryptionBindingV1,
  plaintext: Uint8Array,
  nonce?: Uint8Array,
): AuraEncryptedEnvelopeV1 {
  const envelopeNonce = nonce === undefined
    ? new Uint8Array(randomBytes(ENCRYPTED_ENVELOPE_V1_NONCE_LEN))
    : requireNonceV1(nonce);
  const envelope: AuraEncryptedEnvelopeV1 = {
    version: ENCRYPTED_ENVELOPE_V1_VERSION,
    algorithmId: ENCRYPTED_ENVELOPE_V1_ALGORITHM_ID,
    senderPublicKey: deriveSessionPublicKeyV1(senderSecretKey),
    receiverPublicKey: {
      bytes: cloneBytes32V1(receiverPublicKey.bytes, "receiverPublicKey.bytes"),
    },
    nonce: envelopeNonce,
    aadContextHash: deriveAadContextHashV1(sessionEncryptionContext, stormEncryptionBinding),
    ciphertext: encryptPayloadV1(
      senderSecretKey,
      receiverPublicKey,
      sessionEncryptionContext,
      stormEncryptionBinding,
      envelopeNonce,
      plaintext,
    ),
    sessionKeyId: cloneBytes32V1(sessionEncryptionContext.sessionKeyId, "sessionEncryptionContext.sessionKeyId"),
  };

  return validateEncryptedEnvelopeV1(envelope, sessionEncryptionContext, stormEncryptionBinding);
}

export function validateEncryptedEnvelopeV1(
  envelope: AuraEncryptedEnvelopeV1,
  sessionEncryptionContext: AuraSessionEncryptionContextV1,
  stormEncryptionBinding: StormEncryptionBindingV1,
): AuraEncryptedEnvelopeV1 {
  if (envelope.version !== ENCRYPTED_ENVELOPE_V1_VERSION) {
    throw new TypeError(
      `encrypted envelope version must be ${ENCRYPTED_ENVELOPE_V1_VERSION}, got ${envelope.version}`,
    );
  }

  if (envelope.algorithmId !== ENCRYPTED_ENVELOPE_V1_ALGORITHM_ID) {
    throw new TypeError(
      `encrypted envelope algorithmId must be ${ENCRYPTED_ENVELOPE_V1_ALGORITHM_ID}, got ${envelope.algorithmId}`,
    );
  }

  validateSessionEncryptionContextV1(sessionEncryptionContext);
  validateStormEncryptionBindingV1(stormEncryptionBinding);

  const senderPublicKey = cloneBytes32V1(envelope.senderPublicKey.bytes, "senderPublicKey.bytes");
  const receiverPublicKey = cloneBytes32V1(envelope.receiverPublicKey.bytes, "receiverPublicKey.bytes");
  const sessionKeyId = cloneBytes32V1(envelope.sessionKeyId, "sessionKeyId");
  const contextSessionKeyId = cloneBytes32V1(
    sessionEncryptionContext.sessionKeyId,
    "sessionEncryptionContext.sessionKeyId",
  );
  const bindingSessionKeyId = cloneBytes32V1(
    stormEncryptionBinding.sessionKeyId,
    "stormEncryptionBinding.sessionKeyId",
  );
  const aadContextHash = cloneBytes32V1(envelope.aadContextHash, "aadContextHash");

  if (isAllZeroV1(senderPublicKey)) {
    throw new TypeError("senderPublicKey.bytes must not be all zero");
  }

  if (isAllZeroV1(receiverPublicKey)) {
    throw new TypeError("receiverPublicKey.bytes must not be all zero");
  }

  if (isAllZeroV1(sessionKeyId) || isAllZeroV1(contextSessionKeyId) || isAllZeroV1(bindingSessionKeyId)) {
    throw new TypeError("session key ids must not be all zero");
  }

  if (!equalBytesV1(sessionKeyId, contextSessionKeyId)) {
    throw new TypeError("envelope sessionKeyId does not match the supplied encryption context");
  }

  if (!equalBytesV1(sessionKeyId, bindingSessionKeyId)) {
    throw new TypeError("envelope sessionKeyId does not match the supplied storm encryption binding");
  }

  if (cloneBytesV1(envelope.ciphertext, "ciphertext").length < ENCRYPTED_ENVELOPE_V1_TAG_LEN) {
    throw new TypeError(
      `ciphertext must be at least ${ENCRYPTED_ENVELOPE_V1_TAG_LEN} bytes to include the AEAD tag`,
    );
  }

  const expectedAadContextHash = deriveAadContextHashV1(
    sessionEncryptionContext,
    stormEncryptionBinding,
  );
  if (!equalBytesV1(aadContextHash, expectedAadContextHash)) {
    throw new TypeError("aadContextHash does not match the supplied encryption context and storm binding");
  }

  return {
    version: envelope.version,
    algorithmId: envelope.algorithmId,
    senderPublicKey: { bytes: senderPublicKey },
    receiverPublicKey: { bytes: receiverPublicKey },
    nonce: requireNonceV1(envelope.nonce),
    aadContextHash,
    ciphertext: cloneBytesV1(envelope.ciphertext, "ciphertext"),
    sessionKeyId,
  };
}

export function deriveAadContextHashV1(
  sessionEncryptionContext: AuraSessionEncryptionContextV1,
  stormEncryptionBinding: StormEncryptionBindingV1,
): Uint8Array {
  const aadMaterial = buildAadMaterialV1(sessionEncryptionContext, stormEncryptionBinding);
  return new Uint8Array(
    createHash("sha256")
      .update(AURA_SESSION_ENCRYPTION_AAD_CONTEXT_HASH_V1_DOMAIN_SEPARATOR)
      .update(aadMaterial)
      .digest(),
  );
}

function buildAadMaterialV1(
  sessionEncryptionContext: AuraSessionEncryptionContextV1,
  stormEncryptionBinding: StormEncryptionBindingV1,
): Uint8Array {
  return concatBytesV1(
    encodeSessionEncryptionContextV1(sessionEncryptionContext),
    encodeStormEncryptionBindingV1(stormEncryptionBinding),
  );
}

function requireNonceV1(nonce: Uint8Array): Uint8Array {
  if (!(nonce instanceof Uint8Array) || nonce.length !== ENCRYPTED_ENVELOPE_V1_NONCE_LEN) {
    throw new TypeError(`nonce must be ${ENCRYPTED_ENVELOPE_V1_NONCE_LEN} bytes`);
  }
  return new Uint8Array(nonce);
}

function cloneBytes32V1(bytes: Uint8Array, fieldName: string): Uint8Array {
  if (!(bytes instanceof Uint8Array) || bytes.length !== 32) {
    throw new TypeError(`${fieldName} must be 32 bytes`);
  }
  return new Uint8Array(bytes);
}

function cloneBytesV1(bytes: Uint8Array, fieldName: string): Uint8Array {
  if (!(bytes instanceof Uint8Array)) {
    throw new TypeError(`${fieldName} must be bytes`);
  }
  return new Uint8Array(bytes);
}

function equalBytesV1(left: Uint8Array, right: Uint8Array): boolean {
  if (left.length !== right.length) {
    return false;
  }
  for (let index = 0; index < left.length; index += 1) {
    if ((left[index] ?? 0) !== (right[index] ?? 0)) {
      return false;
    }
  }
  return true;
}

function isAllZeroV1(bytes: Uint8Array): boolean {
  for (const byte of bytes) {
    if (byte !== 0) {
      return false;
    }
  }
  return true;
}

function concatBytesV1(...parts: Uint8Array[]): Uint8Array {
  const totalLength = parts.reduce((sum, part) => sum + part.length, 0);
  const output = new Uint8Array(totalLength);
  let offset = 0;
  for (const part of parts) {
    output.set(part, offset);
    offset += part.length;
  }
  return output;
}

File: crates/aura_intent_lineage_v1/src/symmetric_envelope_v1.rs
//! Aura Session Encryption V1 symmetric envelope construction.

use core::fmt;

use ring::{
    aead::{self, Aad, LessSafeKey, Nonce, UnboundKey},
    rand::{SecureRandom, SystemRandom},
};

use crate::{
    derive_session_public_key_v1, derive_session_symmetric_key_v1, derive_shared_secret_v1,
    encode_session_encryption_context_v1, encode_storm_encryption_binding_v1,
    sha256_domain_separated, validate_session_encryption_context_v1,
    validate_storm_encryption_binding_v1, AuraSessionEncryptionContextV1,
    SessionKeyDerivationInputV1, SessionKeyV1Error, SessionPublicKeyV1, SessionSecretKeyV1,
    StormEncryptionBindingV1, HASH_LEN_V1,
};

pub const ENCRYPTED_ENVELOPE_V1_VERSION: u8 = 0x01;
pub const ENCRYPTED_ENVELOPE_V1_ALGORITHM_ID: u8 = 0x01;
pub const ENCRYPTED_ENVELOPE_V1_NONCE_LEN: usize = 12;
pub const ENCRYPTED_ENVELOPE_V1_TAG_LEN: usize = 16;
pub const AURA_SESSION_ENCRYPTION_AAD_CONTEXT_HASH_V1_DOMAIN_SEPARATOR: &[u8] =
    b"AURA_SESSION_ENCRYPTION_AAD_CONTEXT_HASH_V1";

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct AuraEncryptedEnvelopeV1 {
    pub version: u8,
    pub algorithm_id: u8,
    pub sender_public_key: SessionPublicKeyV1,
    pub receiver_public_key: SessionPublicKeyV1,
    pub nonce: [u8; ENCRYPTED_ENVELOPE_V1_NONCE_LEN],
    pub aad_context_hash: [u8; HASH_LEN_V1],
    pub ciphertext: Vec<u8>,
    pub session_key_id: [u8; HASH_LEN_V1],
}

#[derive(Debug, PartialEq, Eq)]
pub enum SymmetricEnvelopeErrorV1 {
    InvalidEnvelopeVersion {
        expected: u8,
        actual: u8,
    },
    UnsupportedAlgorithm {
        expected: u8,
        actual: u8,
    },
    InvalidSenderPublicKey,
    InvalidReceiverPublicKey,
    InvalidSessionKeyId,
    InvalidCiphertextLength {
        minimum: usize,
        actual: usize,
    },
    AadContextHashMismatch {
        expected: [u8; HASH_LEN_V1],
        actual: [u8; HASH_LEN_V1],
    },
    SessionKeyIdMismatch {
        expected: [u8; HASH_LEN_V1],
        actual: [u8; HASH_LEN_V1],
    },
    InvalidSessionEncryptionContext(crate::SessionEncryptionContextErrorV1),
    InvalidStormEncryptionBinding(crate::StormEncryptionBindingErrorV1),
    InvalidSessionKey(SessionKeyV1Error),
    SecureRandomUnavailable,
    EncryptFailed,
    DecryptFailed,
}

impl fmt::Display for SymmetricEnvelopeErrorV1 {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidEnvelopeVersion { expected, actual } => write!(
                f,
                "invalid encrypted envelope version: expected {expected}, got {actual}"
            ),
            Self::UnsupportedAlgorithm { expected, actual } => write!(
                f,
                "unsupported encrypted envelope algorithm: expected {expected}, got {actual}"
            ),
            Self::InvalidSenderPublicKey => write!(f, "invalid sender public key"),
            Self::InvalidReceiverPublicKey => write!(f, "invalid receiver public key"),
            Self::InvalidSessionKeyId => write!(f, "invalid session key id"),
            Self::InvalidCiphertextLength { minimum, actual } => write!(
                f,
                "invalid ciphertext length: expected at least {minimum} bytes, got {actual}"
            ),
            Self::AadContextHashMismatch { expected, actual } => write!(
                f,
                "aad context hash mismatch: expected {}, got {}",
                crate::LowerHex32(expected),
                crate::LowerHex32(actual)
            ),
            Self::SessionKeyIdMismatch { expected, actual } => write!(
                f,
                "session key id mismatch: expected {}, got {}",
                crate::LowerHex32(expected),
                crate::LowerHex32(actual)
            ),
            Self::InvalidSessionEncryptionContext(error) => {
                write!(f, "invalid session encryption context: {error}")
            }
            Self::InvalidStormEncryptionBinding(error) => {
                write!(f, "invalid storm encryption binding: {error}")
            }
            Self::InvalidSessionKey(error) => write!(f, "invalid session key: {error}"),
            Self::SecureRandomUnavailable => write!(f, "secure randomness unavailable"),
            Self::EncryptFailed => write!(f, "payload encryption failed"),
            Self::DecryptFailed => write!(f, "payload decryption failed"),
        }
    }
}

impl std::error::Error for SymmetricEnvelopeErrorV1 {}

pub fn encrypt_payload_v1(
    sender_secret_key: &SessionSecretKeyV1,
    receiver_public_key: &SessionPublicKeyV1,
    session_encryption_context: &AuraSessionEncryptionContextV1,
    storm_encryption_binding: &StormEncryptionBindingV1,
    nonce: [u8; ENCRYPTED_ENVELOPE_V1_NONCE_LEN],
    plaintext: &[u8],
) -> Result<Vec<u8>, SymmetricEnvelopeErrorV1> {
    let shared_secret = derive_shared_secret_v1(sender_secret_key, receiver_public_key)
        .map_err(SymmetricEnvelopeErrorV1::InvalidSessionKey)?;
    let session_key = derive_session_symmetric_key_v1(&SessionKeyDerivationInputV1 {
        shared_secret,
        session_encryption_context: *session_encryption_context,
        storm_encryption_binding: *storm_encryption_binding,
    })
    .map_err(SymmetricEnvelopeErrorV1::InvalidSessionKey)?;
    let aad_material = build_aad_material_v1(session_encryption_context, storm_encryption_binding)?;

    let unbound_key = UnboundKey::new(&aead::CHACHA20_POLY1305, session_key.as_bytes())
        .map_err(|_| SymmetricEnvelopeErrorV1::EncryptFailed)?;
    let sealing_key = LessSafeKey::new(unbound_key);
    let mut in_out = plaintext.to_vec();
    sealing_key
        .seal_in_place_append_tag(
            Nonce::assume_unique_for_key(nonce),
            Aad::from(aad_material.as_slice()),
            &mut in_out,
        )
        .map_err(|_| SymmetricEnvelopeErrorV1::EncryptFailed)?;
    Ok(in_out)
}

pub fn decrypt_payload_v1(
    receiver_secret_key: &SessionSecretKeyV1,
    sender_public_key: &SessionPublicKeyV1,
    session_encryption_context: &AuraSessionEncryptionContextV1,
    storm_encryption_binding: &StormEncryptionBindingV1,
    nonce: [u8; ENCRYPTED_ENVELOPE_V1_NONCE_LEN],
    ciphertext: &[u8],
) -> Result<Vec<u8>, SymmetricEnvelopeErrorV1> {
    if ciphertext.len() < ENCRYPTED_ENVELOPE_V1_TAG_LEN {
        return Err(SymmetricEnvelopeErrorV1::InvalidCiphertextLength {
            minimum: ENCRYPTED_ENVELOPE_V1_TAG_LEN,
            actual: ciphertext.len(),
        });
    }

    let shared_secret = derive_shared_secret_v1(receiver_secret_key, sender_public_key)
        .map_err(SymmetricEnvelopeErrorV1::InvalidSessionKey)?;
    let session_key = derive_session_symmetric_key_v1(&SessionKeyDerivationInputV1 {
        shared_secret,
        session_encryption_context: *session_encryption_context,
        storm_encryption_binding: *storm_encryption_binding,
    })
    .map_err(SymmetricEnvelopeErrorV1::InvalidSessionKey)?;
    let aad_material = build_aad_material_v1(session_encryption_context, storm_encryption_binding)?;

    let unbound_key = UnboundKey::new(&aead::CHACHA20_POLY1305, session_key.as_bytes())
        .map_err(|_| SymmetricEnvelopeErrorV1::DecryptFailed)?;
    let opening_key = LessSafeKey::new(unbound_key);
    let mut in_out = ciphertext.to_vec();
    let plaintext = opening_key
        .open_in_place(
            Nonce::assume_unique_for_key(nonce),
            Aad::from(aad_material.as_slice()),
            &mut in_out,
        )
        .map_err(|_| SymmetricEnvelopeErrorV1::DecryptFailed)?;

    Ok(plaintext.to_vec())
}

pub fn build_encrypted_envelope_v1(
    sender_secret_key: &SessionSecretKeyV1,
    receiver_public_key: &SessionPublicKeyV1,
    session_encryption_context: &AuraSessionEncryptionContextV1,
    storm_encryption_binding: &StormEncryptionBindingV1,
    plaintext: &[u8],
    nonce: Option<[u8; ENCRYPTED_ENVELOPE_V1_NONCE_LEN]>,
) -> Result<AuraEncryptedEnvelopeV1, SymmetricEnvelopeErrorV1> {
    let sender_public_key = derive_session_public_key_v1(sender_secret_key);
    let nonce = match nonce {
        Some(nonce) => nonce,
        None => generate_envelope_nonce_v1()?,
    };
    let ciphertext = encrypt_payload_v1(
        sender_secret_key,
        receiver_public_key,
        session_encryption_context,
        storm_encryption_binding,
        nonce,
        plaintext,
    )?;
    let envelope = AuraEncryptedEnvelopeV1 {
        version: ENCRYPTED_ENVELOPE_V1_VERSION,
        algorithm_id: ENCRYPTED_ENVELOPE_V1_ALGORITHM_ID,
        sender_public_key,
        receiver_public_key: *receiver_public_key,
        nonce,
        aad_context_hash: derive_aad_context_hash_v1(
            session_encryption_context,
            storm_encryption_binding,
        )?,
        ciphertext,
        session_key_id: session_encryption_context.session_key_id,
    };
    validate_encrypted_envelope_v1(
        &envelope,
        session_encryption_context,
        storm_encryption_binding,
    )?;
    Ok(envelope)
}

pub fn validate_encrypted_envelope_v1(
    envelope: &AuraEncryptedEnvelopeV1,
    session_encryption_context: &AuraSessionEncryptionContextV1,
    storm_encryption_binding: &StormEncryptionBindingV1,
) -> Result<(), SymmetricEnvelopeErrorV1> {
    if envelope.version != ENCRYPTED_ENVELOPE_V1_VERSION {
        return Err(SymmetricEnvelopeErrorV1::InvalidEnvelopeVersion {
            expected: ENCRYPTED_ENVELOPE_V1_VERSION,
            actual: envelope.version,
        });
    }

    if envelope.algorithm_id != ENCRYPTED_ENVELOPE_V1_ALGORITHM_ID {
        return Err(SymmetricEnvelopeErrorV1::UnsupportedAlgorithm {
            expected: ENCRYPTED_ENVELOPE_V1_ALGORITHM_ID,
            actual: envelope.algorithm_id,
        });
    }

    if envelope
        .sender_public_key
        .bytes
        .iter()
        .all(|byte| *byte == 0)
    {
        return Err(SymmetricEnvelopeErrorV1::InvalidSenderPublicKey);
    }

    if envelope
        .receiver_public_key
        .bytes
        .iter()
        .all(|byte| *byte == 0)
    {
        return Err(SymmetricEnvelopeErrorV1::InvalidReceiverPublicKey);
    }

    if session_encryption_context
        .session_key_id
        .iter()
        .all(|byte| *byte == 0)
        || storm_encryption_binding
            .session_key_id
            .iter()
            .all(|byte| *byte == 0)
        || envelope.session_key_id.iter().all(|byte| *byte == 0)
    {
        return Err(SymmetricEnvelopeErrorV1::InvalidSessionKeyId);
    }

    if envelope.ciphertext.len() < ENCRYPTED_ENVELOPE_V1_TAG_LEN {
        return Err(SymmetricEnvelopeErrorV1::InvalidCiphertextLength {
            minimum: ENCRYPTED_ENVELOPE_V1_TAG_LEN,
            actual: envelope.ciphertext.len(),
        });
    }

    validate_session_encryption_context_v1(session_encryption_context)
        .map_err(SymmetricEnvelopeErrorV1::InvalidSessionEncryptionContext)?;
    validate_storm_encryption_binding_v1(storm_encryption_binding)
        .map_err(SymmetricEnvelopeErrorV1::InvalidStormEncryptionBinding)?;

    if envelope.session_key_id != session_encryption_context.session_key_id {
        return Err(SymmetricEnvelopeErrorV1::SessionKeyIdMismatch {
            expected: session_encryption_context.session_key_id,
            actual: envelope.session_key_id,
        });
    }

    if envelope.session_key_id != storm_encryption_binding.session_key_id {
        return Err(SymmetricEnvelopeErrorV1::SessionKeyIdMismatch {
            expected: storm_encryption_binding.session_key_id,
            actual: envelope.session_key_id,
        });
    }

    let expected_aad_context_hash =
        derive_aad_context_hash_v1(session_encryption_context, storm_encryption_binding)?;
    if envelope.aad_context_hash != expected_aad_context_hash {
        return Err(SymmetricEnvelopeErrorV1::AadContextHashMismatch {
            expected: expected_aad_context_hash,
            actual: envelope.aad_context_hash,
        });
    }

    Ok(())
}

pub fn derive_aad_context_hash_v1(
    session_encryption_context: &AuraSessionEncryptionContextV1,
    storm_encryption_binding: &StormEncryptionBindingV1,
) -> Result<[u8; HASH_LEN_V1], SymmetricEnvelopeErrorV1> {
    let aad_material = build_aad_material_v1(session_encryption_context, storm_encryption_binding)?;
    Ok(sha256_domain_separated(
        AURA_SESSION_ENCRYPTION_AAD_CONTEXT_HASH_V1_DOMAIN_SEPARATOR,
        &aad_material,
    ))
}

fn generate_envelope_nonce_v1(
) -> Result<[u8; ENCRYPTED_ENVELOPE_V1_NONCE_LEN], SymmetricEnvelopeErrorV1> {
    let rng = SystemRandom::new();
    let mut nonce = [0u8; ENCRYPTED_ENVELOPE_V1_NONCE_LEN];
    rng.fill(&mut nonce)
        .map_err(|_| SymmetricEnvelopeErrorV1::SecureRandomUnavailable)?;
    Ok(nonce)
}

fn build_aad_material_v1(
    session_encryption_context: &AuraSessionEncryptionContextV1,
    storm_encryption_binding: &StormEncryptionBindingV1,
) -> Result<Vec<u8>, SymmetricEnvelopeErrorV1> {
    let encoded_context = encode_session_encryption_context_v1(session_encryption_context)
        .map_err(SymmetricEnvelopeErrorV1::InvalidSessionEncryptionContext)?;
    let encoded_binding = encode_storm_encryption_binding_v1(storm_encryption_binding)
        .map_err(SymmetricEnvelopeErrorV1::InvalidStormEncryptionBinding)?;
    let mut aad = Vec::with_capacity(encoded_context.len() + encoded_binding.len());
    aad.extend_from_slice(&encoded_context);
    aad.extend_from_slice(&encoded_binding);
    Ok(aad)
}
